### DISCUSSION TITLE
Questions while preparing the test layers

### DISCUSSION THREAD

**Discussion by jmargutt** [rodekruis/ADA-API#32] [General]
POST api/events:
1. where do I find the list of valid values for field "type"?

POST api/events/{id}/layers/{name}
1. what do we do when we want to visualize 2 damage levels instead of 4?
2. for all layers I get this error when uploading, need to specify the schema
<Response [400]> {'message': ['property type should not exist', 'property features should not exist', 'geojson should not be empty', 'information should not be empty', 'information must be a string']}

**Comment by jmargutt** [comment]
@gulfaraz

**Comment by gulfaraz** [comment]
> 1. where do I find the list of valid values for field "type"?

1. In the API swagger documentation for [POST /api/events](https://ada.510.global/api/swagger/#/event/EventController_create),
    - Click **Schema** next to **Example Value**
    - Expand the value for **type**, there we can see an **Enum** array with the valid Event Type values.
      ![screenshot-ada 510 global-2023 02 13-14_33_03](https://user-images.githubusercontent.com/3880591/218728388-b7104fd0-43e0-4b73-a870-3f6e3bae8722.png)
    - For quick reference, the values are,
        - Tropical Cyclone
        - Conflict
        - Eruption
        - Earthquake
        - Fire
        - Flood
        - Heavy Rain
        - Landslide
        - Tsunami
    - The truest source is to look the code in [src/event/event-type.enum.ts](https://github.com/rodekruis/ADA-API/blob/main/src/event/event-type.enum.ts)

> 1. what do we do when we want to visualize 2 damage levels instead of 4?

The layers **buildings-damage-light**, **buildings-damage-moderate**, **buildings-damage-heavy** represent 3 levels of building damage.
The layer **buildings-damage-none** represents no building damage.

These layers are independent of each other in terms of geojson and visualization.
We are free to choose any, none, or all of the above layers to represent building damage in an event.

Below are examples of possible combinations,
3 levels of building damage (by using all layers): **buildings-damage-light**, **buildings-damage-moderate**, **buildings-damage-heavy**
2 levels of building damage (by using a subset of 2): **buildings-damage-light**, **buildings-damage-heavy**
1 level of building damage (by using a subset of 1): **buildings-damage-moderate**

Of course, other combinations are allowed. If the concern is that the level of damage is unclear. Then we can use the `information` field to display an explanation in the layer popup. The `information` field supports markdown so it's possible to add external links for detailed sources and also to include images for clarity.

> 2. for all layers I get this error when uploading, need to specify the schema

[POST /api/events/{id}/layers/{name}](https://ada.510.global/api/swagger/#/event-layer/EventController_createLayer) expects a request body to be in the format,
```json
{
  "geojson": {
    "type": "Feature",
    "geometry": { "type": "Point", "coordinates": [ 125.6, 10.1 ] },
    "properties": { "name": "Dinagat Islands" }
  },
  "information": "Supporting information about this event layer."
}
```

but the requests made by _upload_event_layers.py_ send a request body in the format,
```json
{
  "type": "Feature",
  "geometry": { "type": "Point", "coordinates": [ 125.6, 10.1 ] },
  "properties": { "name": "Dinagat Islands" }
}
```

The ADA-API looks for a request body with keys `geojson` and `information` but finds `type`, `geometry`, and `properties`. Hence throwing the appropriate error messages.

I was able to create the layers using the code below to replace **lines 24-29** in _upload_event_layers.py_,

```python
layer = layer_file.replace('.geojson', '')
layer_data = json.load(f)
request_body = {}
request_body['geojson'] = layer_data
request_body['information'] = f"Popup Information for layer {layer}"
res = requests.post(f'{host}/api/events/{id}/layers/{layer}', headers=headers, data=json.dumps(request_body))
```

@jmargutt

**Comment by gulfaraz** [comment]
I noticed that `admin-1` geojson had features with properties `"name": "ADANA", "number_damaged_buldings": 0.0, "fraction_damaged_buldings": 0.0`

The UI will read the value for `name` but it will ignore `number_damaged_buldings` and `fraction_damaged_buldings`

To fix this, please use,
`building_damage` instead of `number_damaged_buldings`
`building_damage_percentage` instead of `fraction_damaged_buldings`
and `people_affected` if available

@jmargutt

**Comment by jmargutt** [comment]
Hey @gulfaraz , you probably saw that I uploaded some new events over the weekend. Here's my feedback, we can discuss on Thursday but maybe you can start working on something. It's NOT in order of priority, so feel free to pick from easiest to hardest.

1. Zoom in/out buttons and admin level selection disappears from view when "event layers" menu is open. Can we avoid that?
2. Population density layer is barely visible for rural areas (example: Haiti Earthquake), especially with the default zoom level. Can we make it more visible?
3. I did a "stress test" with Cyclone Nargis by uploading a ~50 MB building layer, it is indeed slow to render but not dramatically so. So I'd say about performance we're good.
4. About visualizing different levels of damage (e.g. 2 instead of 4): I saw your answer and I think that's already good. What would be even better (although possibly complicated) is to leave the title of the layer as customizable, if necessary. For example, if I upload only 2 levels of damage, I'd like them to appear as "No / Light Building Damage" and "Moderate / Heavy Building Damage". If too complicated to have the layer name customizable, I'd be happy if you can hard-code this specific scenario, as the 2-level damage assessment is a common thing that we do when we are not super-confident about accuracy. 
5. I would change "shade X based on Building Damage" to "shade X based on Damaged Buildings", to be consistent with the key figures on the left menu.
6. Looking at the analyses that are being requested/produced around the Turkey-Syria earthquake, I do see value in adding two additional ways of "shading" the admin layers: "shade X based on People Affected (%)" and "shade X based on Damaged Buildings (%)". The legends of both layers should follow the same logic as before, 5 intervals equally spaced between 0 and the maximum, plus of course a % symbol to indicate the unit.

**Reply by gulfaraz** [reply]
@jmargutt Thank you for testing and the feedback. Let's review the changes below in our call tomorrow before we merge to the main branch.

> 1. Zoom in/out buttons and admin level selection disappears from view when "event layers" menu is open. Can we avoid that?

Yes, we can set the menu to [_overlay_](https://ionicframework.com/docs/api/menu#type) on the map instead of [_reveal_](https://ionicframework.com/docs/api/menu#type). However, this will add a **translucent backdrop** over the map thus blocking any interaction with the map until the menu is closed.

> 2. Population density layer is barely visible for rural areas (example: Haiti Earthquake), especially with the default zoom level. Can we make it more visible?

https://github.com/rodekruis/ADA-UI/commit/d21620b39fc1eaf87db658e673e63514b78af5d6 improves visiblity of the population density layer. Here is the screenshot with the change,

<img width="953" alt="Screenshot 2023-02-22 at 18 23 49" src="https://user-images.githubusercontent.com/3880591/220707082-4ce73ff0-8d3d-4815-b0ca-1d622dc2ad46.png">

> 3. I did a "stress test" with Cyclone Nargis by uploading a ~50 MB building layer, it is indeed slow to render but not dramatically so. So I'd say about performance we're good.

:tada: Glad to read this. Thank you for testing the performance. :clap:

> 4. About visualizing different levels of damage (e.g. 2 instead of 4): I saw your answer and I think that's already good. What would be even better (although possibly complicated) is to leave the title of the layer as customizable, if necessary. For example, if I upload only 2 levels of damage, I'd like them to appear as "No / Light Building Damage" and "Moderate / Heavy Building Damage". If too complicated to have the layer name customizable, I'd be happy if you can hard-code this specific scenario, as the 2-level damage assessment is a common thing that we do when we are not super-confident about accuracy.

https://github.com/rodekruis/ADA-UI/commit/a37d58cb7b0abf1192fa85441d1baa3cefe9b049 adds support for 2 new layers,
1. `buildings-damage-none-and-light` for **No / Light Damage**
2. `buildings-damage-moderate-and-heavy` for **Moderate / Heavy Damage**

These new layers can be used similarly to the `buildings-damage-x` layers.

> 5. I would change "shade X based on Building Damage" to "shade X based on Damaged Buildings", to be consistent with the key figures on the left menu.

As we use **Building Damage** more often than **Damaged Buildings**, the key figures on the left menu is renamed to **Building Damage**. https://github.com/rodekruis/ADA-UI/commit/6f99741cb443c0757cf5edac3f2d43a8efecf0c2 uses **Building Damage** consistently throughout the app.

> 6. Looking at the analyses that are being requested/produced around the Turkey-Syria earthquake, I do see value in adding two additional ways of "shading" the admin layers: "shade X based on People Affected (%)" and "shade X based on Damaged Buildings (%)". The legends of both layers should follow the same logic as before, 5 intervals equally spaced between 0 and the maximum, plus of course a % symbol to indicate the unit.

https://github.com/rodekruis/ADA-UI/commit/848f4161529246d6a00789d0615e8669a5933ebd adds "shade X based on People Affected (%)" and "shade X based on Building Damage (%)" for _shading_ the admin layers. It also adds these to the admin area popups. And adds the missing value to the key figures on the left menu.

The geojson properties used for these will be `people_affected_percentage` and `building_damage_percentage`.
